### PR TITLE
convert kafka stream test from groovy to java

### PR DESCRIPTION
**Pull request by shalk**
Related to #7195

### CODE REVIEW

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsDefaultTest.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+  @Test
+  void testKafkaProduceAndConsumeWithStreamsInBetween()
+      throws ExecutionException,
+          InterruptedException,
+          TimeoutException,
+          InstantiationException,
+          IllegalAccessException,
+          ClassNotFoundException,
+          NoSuchMethodException,
+          InvocationTargetException {
```
Comment (jaydeluca) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsDefaultTest.java:61]: For tests I don't think we need to be so explicit when there's so many, and can clean this up a bit via:

```suggestion
  void testKafkaProduceAndConsumeWithStreamsInBetween() throws Exception {
```

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamReflectionUtil.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    try {
+      // Different class names for test and latestDepTest.
+      return Class.forName("org.apache.kafka.streams.kstream.KStreamBuilder").newInstance();
+    } catch (ClassNotFoundException
+        | NoClassDefFoundError
+        | InstantiationException
+        | IllegalAccessException e) {
+      return Class.forName("org.apache.kafka.streams.StreamsBuilder").newInstance();
+    }
+  }
```
Comment (jaydeluca) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamReflectionUtil.java:36]: perhaps we could avoid the need for the  `@SuppressWarnings("ClassNewInstance")` if we did something like this?

```suggestion
  static Object createBuilder() throws Exception {
    try {
      // Different class names for test and latestDepTest.
      return Class.forName("org.apache.kafka.streams.kstream.KStreamBuilder").getConstructors()[0].newInstance();
    } catch (Exception e) {
      return Class.forName("org.apache.kafka.streams.StreamsBuilder").getConstructors()[0].newInstance();
    }
  }
```

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamReflectionUtil.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+  @SuppressWarnings("unchecked")
+  static KStream<Integer, String> stream(Object builder, String topic)
+      throws ClassNotFoundException,
+          NoSuchMethodException,
+          InvocationTargetException,
+          IllegalAccessException {
```
Comment (jaydeluca) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamReflectionUtil.java:43]: Same consideration for the rest of these in this and other files

```suggestion
  static KStream<Integer, String> stream(Object builder, String topic) throws Exception
```

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsDefaultTest.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    awaitUntilConsumerIsReady();
+    @SuppressWarnings("PreferJavaTimeOverload")
+    ConsumerRecords<Integer, String> records =
+        KafkaStreamsBaseTest.consumer.poll(Duration.ofSeconds(10).toMillis());
+    Headers receivedHeaders = null;
+    for (ConsumerRecord<Integer, String> record : records) {
+      Span.current().setAttribute("testing", 123);
+
+      assertEquals(10, record.key());
+      assertEquals(greeting.toLowerCase(Locale.ROOT), record.value());
```
Comment (jaydeluca) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsDefaultTest.java:95]:
```suggestion
      assertThat(record.key()).isEqualTo(10);
      assertThat(record.value()).isEqualTo(greeting.toLowerCase(Locale.ROOT));
```

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsSuppressReceiveSpansTest.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                                  longKey("kafka.record.queue_time_ms"),
+                                  k -> k.isGreaterThanOrEqualTo(0)),
+                              equalTo(stringKey("asdf"), "testing")));
+                  if (Boolean.getBoolean("testLatestDeps")) {
+                    assertions.add(
+                        equalTo(
+                            MessagingIncubatingAttributes.MESSAGING_KAFKA_CONSUMER_GROUP,
+                            "test-application"));
+                  }
+                  ;
```
Comment (jaydeluca) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsSuppressReceiveSpansTest.java:156]:
```suggestion
```

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsBaseTest.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        "33554432",
+        "key.serializer",
+        IntegerSerializer.class,
+        "value.serializer",
+        StringSerializer.class);
+  }
+
+  // Kafka's eventual consistency behavior forces us to do a couple of empty poll() calls until it
+  // gets properly assigned a topic partition
+  @SuppressWarnings("PreferJavaTimeOverload")
```
Comment (laurit) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsBaseTest.java:133]: why is this suppression necessary?

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsDefaultTest.java [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+            STREAM_PROCESSED + " receive"),
+        trace -> {
+          trace.hasSpansSatisfyingExactly(
+              // kafka-clients PRODUCER
+              span ->
+                  span.hasName(STREAM_PENDING + " publish")
+                      .hasKind(SpanKind.PRODUCER)
+                      .hasNoParent()
+                      .hasAttributesSatisfyingExactly(
+                          equalTo(MessagingIncubatingAttributes.MESSAGING_SYSTEM, "kafka"),
```
Comment (laurit) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsDefaultTest.java:106]: there should be a constant for this in the `MessagingIncubatingAttributes`

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamReflectionUtil.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+ */
+class KafkaStreamReflectionUtil {
+
+  private KafkaStreamReflectionUtil() {}
+
+  static Object createBuilder() throws Exception {
+    try {
+      // Different class names for test and latestDepTest.
+      return Class.forName("org.apache.kafka.streams.kstream.KStreamBuilder")
+          .getConstructors()[0]
```
Comment (laurit) [inline instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamReflectionUtil.java:28]: If you are looking for the no-arg constructor you should call `getConstructor()`. If the class has multiple constructors the order of elements in the array returned by `getConstructors` is unspecified.

File: instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsBaseTest.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+  @BeforeAll
+  static void setup() throws ExecutionException, InterruptedException, TimeoutException {
+    kafka =
+        new KafkaContainer(DockerImageName.parse("apache/kafka:3.8.0"))
+            .withEnv("KAFKA_HEAP_OPTS", "-Xmx256m")
+            .waitingFor(Wait.forLogMessage(".*started \\(kafka.server.Kafka.*Server\\).*", 1))
+            .withStartupTimeout(Duration.ofMinutes(1));
+    kafka.start();
+    cleanup.deferCleanup(kafka::stop);
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/kafka/kafka-streams-0.11/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/kafkastreams/KafkaStreamsBaseTest.java:67]: also `AutoCleanupExtension` executes the actions in the order they were added, here you need to stop the kafka container after consumer and the producer

**Review by laurit — APPROVED** [review]

**Pull request merged**